### PR TITLE
Remove sympy.Tuple from pyccel's AST representation

### LINKED ISSUE #657: Remove sympy.Tuple from Pyccel
`sympy.Tuple` does not always symplify it's arguments, however the `sympify=False` argument is easily forgotten which leads to sympy objects replacing pyccel objects. `tuple` is sufficient in pyccel

### PR DESCRIPTION
**Pull request by EmilyBourne**
This PR removes `sympy.Tuple` from Pyccel's AST representation, because it is not necessary. Where it is used to group items a python `tuple` is sufficient. Where it may represent tuples in the translated code `PythonTuple` should be used instead. Unless `sympy.Tuple` is created with `sympify = False` it also converts its arguments to sympy objects, making it a common source of errors. This fixes #657.

### CODE REVIEW

File: pyccel/ast/builtins.py [hunk 1]
```diff
@@ -402,7 +402,7 @@ def arg(self):
         return self._args[0]
 
 #==============================================================================
-class PythonList(Tuple, PyccelAstNode):
+class PythonList(PythonTuple, PyccelAstNode):
```
Comment (yguclu) [inline pyccel/ast/builtins.py:405]: Is `PythonTuple` not a subclass of `PyccelAstNode`?

Comment (EmilyBourne) [reply pyccel/ast/builtins.py:405]: You're right

File: pyccel/ast/core.py [hunk 2]
```diff
@@ -2783,7 +2782,7 @@ def __init__(self, func, args, current_function=None):
         if str(current_function) == str(name):
             func.set_recursive()
 
-        if not isinstance(args, (tuple, list, Tuple)):
+        if not isinstance(args, (tuple, list)):
             raise TypeError('> expecting an iterable')
```
Comment (yguclu) [inline pyccel/ast/core.py:2786]: Could you please change this error message?
A general iterable object is not accepted here, it must be either `list` or `tuple`.

File: pyccel/ast/core.py [hunk 3]
```diff
@@ -3084,7 +3083,7 @@ def __init__(
             raise TypeError('Expecting a boolean for header')
 
         if arguments_inout:
-            if not isinstance(arguments_inout, (list, tuple, Tuple)):
+            if not isinstance(arguments_inout, (list, tuple)):
                 raise TypeError('Expecting an iterable ')
```
Comment (yguclu) [inline pyccel/ast/core.py:3087]: Same comment as above

File: pyccel/ast/core.py [hunk 4]
```diff
@@ -4081,27 +4080,25 @@ def __new__(
         as_lambda=False,
         nargs=1,
         ):
-        if not isinstance(module, (str, DottedName, list, tuple,
-                          Tuple)):
+        if not isinstance(module, (str, DottedName, list, tuple)):
             raise TypeError('Expecting a string or DottedName, given {0}'.format(type(module)))
```
Comment (yguclu) [inline pyccel/ast/core.py:4084]: This error message is also incomplete. Can we fix it in this PR?

File: pyccel/ast/core.py [hunk 5]
```diff
@@ -4081,27 +4080,25 @@ def __new__(
...
 
-        if isinstance(module, (list, tuple, Tuple)):
+        if isinstance(module, (list, tuple)):
             module = DottedName(*module)
 
         if funcs:
-            if not isinstance(funcs, (str, DottedName, list, tuple,
-                              Tuple)):
+            if not isinstance(funcs, (str, DottedName, list, tuple)):
                 raise TypeError('Expecting a string or DottedName')
```
Comment (yguclu) [inline pyccel/ast/core.py:4096]: Same comment as above

File: pyccel/ast/core.py [hunk 6]
```diff
@@ -4378,7 +4375,7 @@ def __new__(
         iterator,
         stmts=None,
         ):
-        if not isinstance(iterator, (tuple, Tuple)):
+        if not isinstance(iterator, tuple):
             raise TypeError('iterator must be a tuple')
         if not len(iterator) == 3:
             raise ValueError('iterator must be of lenght 3')
```
Comment (yguclu) [inline pyccel/ast/core.py:4381]: Typo: `lenght` -> `length`

File: pyccel/ast/builtins.py [hunk 7]
```diff
@@ -548,7 +548,7 @@ class PythonZip(Basic):
     """
 
     def __new__(cls, *args):
-        if not isinstance(args, (tuple, list, Tuple)):
+        if not isinstance(args, (tuple, list)):
             raise TypeError('args must be an iterable')
```
Comment (yguclu) [inline pyccel/ast/builtins.py:552]: I've just noticed that this error message is also wrong...

File: pyccel/ast/core.py [hunk 8]
```diff
@@ -11,7 +11,7 @@
 from sympy import sympify
 from sympy import Add as sp_Add, Mul as sp_Mul, Pow as sp_Pow
 from sympy import Eq as sp_Eq, Ne as sp_Ne, Lt as sp_Lt, Le as sp_Le, Gt as sp_Gt, Ge as sp_Ge
-from sympy import Integral, Symbol, Tuple
+from sympy import Integral, Symbol
```
Comment (yguclu) [inline pyccel/ast/core.py:14]: I know that this is not strictly related to this PR, but did you know that we were importing `Integral` from `sympy`?? Apparently it only appears in line 286 of this file, where `Derivative` is also used (which is imported on line 24)...

@saidctb Do you know anything about this? Is it used by Psydac?

Comment (saidctb) [reply pyccel/ast/core.py:14]: it is used in `allocatable_like`  but I don't think it's used anywhere we can delete it

Comment (EmilyBourne) [reply pyccel/ast/core.py:14]: This comment is about issue #663. That issue is for all the extra sympy objects which are imported but used infrequently enough to not require their own PR

File: pyccel/parser/semantic.py [hunk 9]
```diff
@@ -9,6 +9,7 @@
 from collections import OrderedDict
 from itertools import chain
 
+from sympy import Tuple
```
Comment (yguclu) [inline pyccel/parser/semantic.py:12]: I'm confused. Isn't the goal of this PR to completely eradicate `sympy.Tuple` from Pyccel?

File: pyccel/complexity/arithmetic.py [hunk 10]
```diff
@@ -63,7 +62,7 @@ def count_ops(expr, visual=None):
         a = expr.iterable.size
         ops = sum(count_ops(i, visual) for i in expr.body.body)
         return a*ops
-    elif isinstance(expr, (Tuple,PythonTuple)):
+    elif isinstance(expr, Tuple):
```
Comment (yguclu) [inline pyccel/complexity/arithmetic.py:65]: Do we still need to use `Tuple` here?

File: pyccel/symbolic/lambdify.py [hunk 11]
```diff
@@ -123,7 +123,7 @@ def lambdify(expr, args):
     return func
 
 def set_fst(expr, fst):
-    if isinstance(expr, (tuple,list,Tuple)):
+    if isinstance(expr, (tuple,list)):
```
Comment (yguclu) [inline pyccel/symbolic/lambdify.py:126]: We are still using `Tuple` at line 58 of this file. Can we get rid of it?

File: pyccel/parser/semantic.py (hunk 9 above)
Comment (EmilyBourne) [reply pyccel/parser/semantic.py:12]: It is used once in this file to create a temporary object in order to use the sympy atoms function. We cannot easily add or own version until we have removed most sympy strangeness. I think it should be possible after this PR.

File: pyccel/ast/core.py (hunk 8 above)
Comment (yguclu) [reply pyccel/ast/core.py:14]: OK, thanks!

File: pyccel/complexity/arithmetic.py (hunk 10 above)
Comment (EmilyBourne) [reply pyccel/complexity/arithmetic.py:65]: The files in this folder begin with a call to pyccel_to_sympy

File: pyccel/parser/semantic.py (hunk 9 above)
Comment (yguclu) [reply pyccel/parser/semantic.py:12]: I see. @saidctb Can you have a quick look into this?

File: pyccel/symbolic/lambdify.py (hunk 11 above)
Comment (EmilyBourne) [reply pyccel/symbolic/lambdify.py:126]: I believe lambdify uses sympy to create a usable temporary function. But I'm not sure

File: pyccel/complexity/arithmetic.py (hunk 10 above)
Comment (yguclu) [reply pyccel/complexity/arithmetic.py:65]: I see, so we need to use Sympy in certain places in order to perform specific operations. This should be clarified in the PR's title and description. Maybe change the title to "Remove sympy.Tuple from Pyccel's AST"?

File: pyccel/symbolic/lambdify.py (hunk 11 above)
Comment (yguclu) [reply pyccel/symbolic/lambdify.py:126]: I think it creates a Pyccel `FunctionDef` starting from `sympy.Lambda` or something else...

File: pyccel/complexity/arithmetic.py (hunk 10 above)
Comment (EmilyBourne) [reply pyccel/complexity/arithmetic.py:65]: I don't think this file is used by pyccel at the moment. Sympy should be kept where it is used as designed, ie to simplify equations etc

File: pyccel/symbolic/lambdify.py (hunk 11 above)
Comment (EmilyBourne) [reply pyccel/symbolic/lambdify.py:126]: I have just had another look at this. You are right, it should be replaced, but I don't know enough about this file to know if it should be replaced with `tuple` or `PythonTuple`. @saidctb do you know?

Comment (saidctb) [reply pyccel/symbolic/lambdify.py:126]: this is for symbolic manipulations, so leave sympy there for the moment

File: pyccel/parser/semantic.py (hunk 9 above)
Comment (saidctb) [reply pyccel/parser/semantic.py:12]: it's only used here, https://github.com/pyccel/pyccel/blob/ee4c3259c42ef0533fedee1f96161df0f45fbb59/pyccel/parser/semantic.py#L2605
which for sure doesn't work properly anymore 
we need to use `_atomic` for that

**Review by yguclu — APPROVED** [review]

File: pyccel/symbolic/lambdify.py (hunk 11 above)
Comment (yguclu) [reply pyccel/symbolic/lambdify.py:126]: OK then, I think we can merge this PR as it is.

File: pyccel/complexity/arithmetic.py (hunk 10 above)
Comment (yguclu) [reply pyccel/complexity/arithmetic.py:65]: Fine with me.

**Review by saidctb — APPROVED** [review]
